File: java/src/Color.java
/******************************************************************************
 *                                  LICENSE                                   *
 ******************************************************************************
 *  This file is part of newtonian_black_holes.                               *
 *                                                                            *
 *  newtonian_black_holes is free software: you can redistribute it and/or    *
 *  modify it under the terms of the GNU General Public License as published  *
 *  by the Free Software Foundation, either version 3 of the License, or      *
 *  (at your option) any later version.                                       *
 *                                                                            *
 *  newtonian_black_holes is distributed in the hope that it will be useful   *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of            *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the             *
 *  GNU General Public License for more details.                              *
 *                                                                            *
 *  You should have received a copy of the GNU General Public License         *
 *  along with newtonian_black_holes.  If not, see                            *
 *  <https://www.gnu.org/licenses/>.                                          *
 ******************************************************************************/
package nbh;

public class Color {
    public int red, green, blue;

    public Color(int r, int g, int b) {
        red = r;
        green = g;
        blue = b;
    }

    public void write(PPM ppm) {
        ppm.write(String.format("%d %d %d%n", red, green, blue));
    }

    public Color scale(double t) {
        int r = (int)(t * red);
        int g = (int)(t * green);
        int b = (int)(t * blue);
        return new Color(r, g, b);
    }

    public void scaleBy(double t) {
        red = (int)(t * red);
        green = (int)(t * green);
        blue = (int)(t * blue);
    }

    public static Color checkerBoard(Vec6 u) {
        int cx, cy, use_white;
        double color_factor = Setup.z_detector_squared / u.p.normSquared();
        int val = (int)(255.0 * color_factor);

        if (u.p.z < Setup.z_detector) {
            return new Color(0, 0, 0);
        }

        cx = (int)(Math.ceil(u.p.x));
        cy = (int)(Math.ceil(u.p.y));
        use_white = (cx + cy) & 1;

        if (use_white != 0) {
            return new Color(val, val, val);
        } else {
            return new Color(val, 0, 0);
        }
    }
}

File: swift/color.swift
/******************************************************************************
 *                                  LICENSE                                   *
 ******************************************************************************
 *  This file is part of newtonian_black_holes.                               *
 *                                                                            *
 *  newtonian_black_holes is free software: you can redistribute it and/or    *
 *  modify it under the terms of the GNU General Public License as published  *
 *  by the Free Software Foundation, either version 3 of the License, or      *
 *  (at your option) any later version.                                       *
 *                                                                            *
 *  newtonian_black_holes is distributed in the hope that it will be useful   *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of            *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the             *
 *  GNU General Public License for more details.                              *
 *                                                                            *
 *  You should have received a copy of the GNU General Public License         *
 *  along with newtonian_black_holes.  If not, see                            *
 *  <https://www.gnu.org/licenses/>.                                          *
 ******************************************************************************/

/*  C input-output routines provided here, like fopen, fclose, and fputs.     */
#if canImport(Darwin)
import Darwin
#elseif canImport(Glibc)
import Glibc
#else
import Foundation
#endif

struct Color {
    var red, green, blue: Int32

    func writeToFile(file: UnsafeMutablePointer<FILE>) {
        putc(red, file)
        putc(green, file)
        putc(blue, file)
    }

    func writeToPPM(ppm: PPM) {
        self.writeToFile(file: ppm.file)
    }
}

extension Color {
    static func + (left: Color, right: Color) -> Color {
        let red: Int32 = Int32(0.5*(Double(left.red) + Double(right.red)))
        let green: Int32 = Int32(0.5*(Double(left.green) + Double(right.green)))
        let blue: Int32 = Int32(0.5*(Double(left.blue) + Double(right.blue)))
        return Color(red: red, green: green, blue: blue)
    }

    static func += (left: inout Color, right: Color) {
        left = left + right
    }

    static func * (left: Color, right: Double) -> Color {
        let red: Int32 = Int32(Double(left.red) * right)
        let green: Int32 = Int32(Double(left.green) * right)
        let blue: Int32 = Int32(Double(left.blue) * right)
        return Color(red: red, green: green, blue: blue)
    }

    static func * (left: Double, right: Color) -> Color {
        let red: Int32 = Int32(left * Double(right.red))
        let green: Int32 = Int32(left * Double(right.green))
        let blue: Int32 = Int32(left * Double(right.blue))
        return Color(red: red, green: green, blue: blue)
    }

    static func *= (left: inout Color, right: Double) {
        left = left * right
    }
}

let black: Color = Color(red:   0, green:   0, blue:   0)
let white: Color = Color(red: 255, green: 255, blue: 255)
let red:   Color = Color(red: 255, green:   0, blue:   0)

func checkerBoard(point: Vec3) -> Color {
    let color_factor = zDetectorSquared / point.normSquared()
    if point.z < zDetector {
        return black
    } else if point.normSquared() <= blackHoleRadiusSquared {
        return black
    } else {
        let cx: Int32 = Int32(ceil(point.x))
        let cy: Int32 = Int32(ceil(point.y))
        let use_white: Int32 = (cx + cy) & 1

        if use_white != 0 {
            return color_factor * white
        } else {
            return color_factor * red
        }
    }
}
